landlab/components/groundwater/dupuit_percolator.py

```diff
@@ -0,0 +1,197 @@
+    >>> from landlab import RasterModelGrid
+    >>> mg = RasterModelGrid((3, 3))
+    >>> z = mg.add_zeros('node', 'topographic__elevation')
+    >>> gdp = GroundwaterDupuitPercolator(mg)
+
+    Notes
+    -----
+    Groundwater discharge per unit length, q, is calculated as:
+
+        q = - K H dw/dx,
+
+    where K is hydraulic conductivity, H is aquifer thickness, w is water table
+    height, and x is horizontal distance.
+
+    An explicit forward-in-time finite-volume method is used to implement a
+    numerical solution. Flow discharge between neighboring nodes is calculated
+    using the average depth at the nodes.
+
+    Note that the current version does NOT handle surface seepage.
+    """
+
+    _name = "GroundwaterDupuitPercolator"
+
+    _input_var_names = set(("topographic__elevation",
+                            "aquifer_base__elevation"))
+
+    _output_var_names = set(
+        ("aquifer__thickness", "water_table__elevation", "hydraulic__gradient",
+         "groundwater__specific_discharge", "groundwater__velocity")
+    )
+
+    _var_units = {
+        "topographic__elevation": "m",
+        "aquifer_base__elevation": "m",
+        "aquifer__thickness": "m",
+        "water_table__elevation": "m",
+        "hydraulic__gradient": "m/m",
+        "groundwater__specific_discharge": "m2/s",
+        "groundwater__velocity": "m/s"
+    }
+
+    _var_mapping = {
+        "topographic__elevation": "node",
+        "aquifer_base__elevation": "node",
+        "aquifer__thickness": "node",
+        "water_table__elevation": "node",
+        "hydraulic__gradient": "link",
+        "groundwater__specific_discharge": "link",
+        "groundwater__velocity": "link",
+    }
+
+    _var_doc = {
+        "topographic__elevation": "elevation of land surface",
+        "aquifer_base__elevation": "elevation of impervious layer",
+        "aquifer__thickness": "thickness of saturated zone",
+        "water_table__elevation": "elevation of water table",
+        "hydraulic__gradient": "gradient of water table in link direction",
+        "groundwater__specific_discharge": "discharge per width in link dir",
+        "groundwater__velocity": "velocity of groundwater in link direction",
+    }
+
+    def __init__(self, grid, hydraulic_conductivity=0.01,
+                 recharge_rate=1.0e-8):
+        """Initialize the GroundwaterDupuitPercolator.
+
+        Parameters
+        ----------
+        grid: ModelGrid
+            Landlab ModelGrid object
+        hydraulic_conductivity: float, field name, or array of float
+                saturated hydraulic conductivity, m/s
+                Default = 0.01 m/s
+        recharge_rate: float, field name, or array of float
+                Rate of recharge, m/s
+                Default = 1.0e-8 m/s
+        """
+        # Store grid
+        self._grid = grid
+
+        # Shorthand
+        self.cores = grid.core_nodes
+        self.inactive_links = np.where(grid.status_at_link != ACTIVE_LINK)[0]
```

Comment:
Make life easier for future you and make all these private attributes?

Reply:
good idea